onboarding/mlsys/2-autodiff.md

```diff
@@ -0,0 +1,399 @@
+output of the previous layer.
+
+following table representing the forward pass (that is, where you
+calculate outputs of each sequential layer).  Let `x = 3`, `ytrue
+= 5`, and `m = 7`.
+
+| layer                 |  `f(n)`        | output | `f'(n)` | `f'(g(x))` |
+| --------------------- | -------------- | ------ | ------- | ---------- |
+| start with $x = 3$    |   n/a          |   1    |   n/a   |    n/a     |
+| multiply by $m = 7$   |  $m \cdot n$   |        |   $m$   |            |
+| subtract $ytrue = 5$  |  $n-5$         |        |   $1$   |            |
+| square the result     |  $n^2$         |        |   $2n$  |            |
+
+And lastly, what is the product of all the numbers in the final column?
+
+### A Complete Example, In PyTorch
+
+Now let's have PyTorch do the same thing for us.
+
+Start with some tensors `x` and `y`.
+
+```python
+import torch
+x = torch.tensor([3.0], requires_grad=True)
+y = torch.tensor([5.0], requires_grad=True)
+```
+
+Then create a trainable tensor `m`.  "Trainable" means that we want PyTorch
+to calculate gradients during backpropagation, which we can use to modify the
+tensor.
+
+```python
+m = torch.tensor([7.0], requires_grad=True)
+```
+
+Now do a manual forward pass, feeding the output of each layer to the input of
+the next, and print the results.  Then ask PyTorch to do a backward pass by
+calling `loss.backward()`, which will set an `m.grad` variable representing the
+gradient for `m`.
+
+```python
+print("-- forward pass --")
+mx = x * m
+err = mx - y
+loss = err.square()
+
+print("m initial:", m)
+print("mx forward:", mx)
+print("err forward:", err)
+print("loss forward:", loss)
+
+#
+# Q2 code will go here in just a bit
+#
+
+print("-- backward pass --")
+loss.backward()
+print(f"-- final gradient: {m.grad} --")
+```
+
+Make sure that the forward pass numbers match the output column you calculated
+in Q1, and the final gradient matches the product of the numbers in your final
+column from Q1.
+
+### Q2
+
+Write code to trace the calculation of `m.grad` by adding backward hooks to
+each of the following tensors:
+
+- `m`
+- `mx`
+- `err`
+- `loss`
+
+Use this as an excuse to get familiar with PyTorch internals if you aren't
+already familiar.  I suggest checking out the PyTorch codebase to your machine,
+then browsing via your IDE or `git grep` to find where the `torch.Tensor` class
+is defined, then looking around in there.
+
+What order do the backward hooks print, and what are the gradients at each
+step?  Explain what you see happening.
+
+### Q3
+
+If you rerun your forward pass (don't reinitialize `m`, but start with
+`mx = x * m`), then rerun `loss.backward()`, what effect does it have on your
+`m.grad`?
+
+Fun fact: multiple forward passes and `backward()` calls without zeroing
+gradients in between is exactly how we implement gradient aggregation in
+`PyTorchTrial` (only we don't reuse the same data for both forward passes).
```

Comment:
Maybe worth stating that this behavior is a happy side effect of the need for gradients to sum in this way in order to compute backpropagation correctly in the first place. The summation is needed in cases where a parameter is used in multiple operations.

Might be too detailed, though.